core/src/main/java/com/linecorp/armeria/client/ProxyBuilder.java

```diff
@@ -0,0 +1,177 @@
+
+    @Nullable
+    protected String getPassword() {
+        return password;
+    }
+
+    protected void setPassword(@Nullable String password) {
+        this.password = password;
+    }
+
+    protected InetSocketAddress getProxyAddress() {
+        return proxyAddress;
+    }
+
+    protected long getConnectTimeoutMillis() {
+        return connectTimeoutMillis;
+    }
+
+    protected void setUseSsl(boolean useSsl) {
+        this.useSsl = useSsl;
+    }
+
+    protected boolean getUseSsl() {
+        return useSsl;
+    }
+
+    /**
+     * Build the proxy object.
+     */
+    public abstract Proxy build();
+
+    /**
+     * A {@code ProxyBuilder} which builds a SOCKS4 protocol configuration.
+     */
+    public static final class Socks4ProxyBuilder extends ProxyBuilder {
+        Socks4ProxyBuilder(InetSocketAddress proxyAddress, long connectTimeoutMillis) {
+            super(proxyAddress, connectTimeoutMillis);
+        }
+
+        /**
+         * Set the proxy userName.
+         */
+        public Socks4ProxyBuilder userName(String userName) {
+            setUserName(userName);
+            return this;
+        }
+
+        @Override
+        public Socks4Proxy build() {
+            final Socks4Proxy socks4Proxy =
+                    new Socks4Proxy(getProxyAddress(), getConnectTimeoutMillis());
+            socks4Proxy.setUserName(getUserName());
+            return socks4Proxy;
+        }
+    }
+
+    /**
+     * A {@code ProxyBuilder} which builds a SOCKS4 protocol configuration.
+     */
+    public static final class Socks5ProxyBuilder extends ProxyBuilder {
+        Socks5ProxyBuilder(InetSocketAddress proxyAddress, long connectTimeoutMillis) {
+            super(proxyAddress, connectTimeoutMillis);
+        }
+
+        /**
+         * Set the proxy userName.
+         */
+        public Socks5ProxyBuilder userName(String userName) {
+            setUserName(userName);
+            return this;
+        }
+
+        /**
+         * Set the proxy password.
+         */
+        public Socks5ProxyBuilder password(String password) {
+            setPassword(password);
+            return this;
+        }
+
+        @Override
+        public Socks5Proxy build() {
+            final Socks5Proxy socks5Proxy =
+                    new Socks5Proxy(getProxyAddress(), getConnectTimeoutMillis());
+            socks5Proxy.setUserName(getUserName());
+            socks5Proxy.setPassword(getPassword());
+            return socks5Proxy;
+        }
+    }
+
+    /**
+     * A {@code ProxyBuilder} which builds a SOCKS4 protocol configuration.
+     */
+    public static final class ConnectProxyBuilder extends ProxyBuilder {
+        ConnectProxyBuilder(InetSocketAddress proxyAddress, long connectTimeoutMillis) {
+            super(proxyAddress, connectTimeoutMillis);
+        }
+
+        /**
+         * Set the proxy userName and password.
+         */
+        public ConnectProxyBuilder auth(String userName, String password) {
```

Comment:
How about just splitting into `username()` and `password()`?

Reply:
Actually, as you may know `HttpProxyHandler` requires both `username`, `password` are either null or non-null.

If we decide to split this, I can add validation logic to the `build` method, but I felt this style of API would communicate this restraint to the user better. Let me know what you think 😉

Reply:
OK, then how about renaming to `credential()`?